### ISSUE TITLE
GUI installer for AppImageLauncherLite

### ISSUE THREAD

**Issue by probonopd** [TheAssassin/AppImageLauncher#243]
When the AppImageLauncherLite AppImage is double-clicked, a GUI installer should check whether AppImageLauncherLite is correctly installed in the system and if yes, launch it using systemd; if no, install it.

**Comment by TheAssassin** [comment]
I guess the whole point of AppImageLauncher Lite is that you need the CLI once, but only once. But don't get me wrong, I like this idea. Not a priority for me, but PRs highly welcome.

**Comment by probonopd** [comment]
After thinking about it some more, I think the best cause of action is to actually embed some dialogs into the existing `AppRun` bash script. Maybe using [qarma](https://github.com/luebking/qarma)?

**Comment by TheAssassin** [comment]
An alternative is to make a GUI in C++ with Qt. I don't care however it's done, as long as the UX is good.

Bloating the existing script is however a bad idea. Just make a separate one, and launch that if there's no arguments passed and e.g., `$DISPLAY` is not set. That way you can also test it easily.

**Comment by probonopd** [comment]
That was my original thought, but the more I think about it I believe the best way is to wrap each and every step in proper error handling with GUI, hence to combine the logic that does stuff with checking error codes and using the GUI to react to them.

**Comment by TheAssassin** [comment]
Err... *what*?

**Comment by probonopd** [comment]
Pseudo-code:

```
test_someting
if $RESULT != FOO ; then qarma "Question..." ; fi
do_something
if $? != 0 ; then qarma "Error..." ; fi
```

**Comment by TheAssassin** [comment]
I still think it makes sense to move any form of GUI into a second script. What you suggest also seems very verbose and hard to debug. Therefore separating it makes sense, too.

**Comment by probonopd** [comment]
Do you realize that it will be orders of magnitude more work that way unless I am missing something?

**Comment by probonopd** [comment]
> The easiest way I can think of to allow for providing both CLI and UI would be to extract the install_ methods into a sourceable script, source that from the respective UI/CLI installers and then call the install steps one by one with appropriate error messages

Yes, that'd be a good way of doing it.

> In fact I'm tempted to get rid of the entire shell script and code the whole CLI installer in C++. 

That as well ;-)

**Comment by TheAssassin** [comment]
Prepared everything for you to get started. You can start off by copying the existing CLI AppRun script and modifying it to display UI messages. The "interesting" bits have been extracted into a `source`able script.

**Comment by probonopd** [comment]
> Prepared everything for you to get started.

Can you give me a "hello world" in terms of what I should do? Thanks. (I better ask before I do it wrong.)

**Comment by TheAssassin** [comment]
You had some plan of patching the script by adding tons of `|| my_ui_tool "Error: abcdefg"`, hadn't you? You could just copy the existing installer script and append those messages behind every install/uninstall function call.

Also, you need to find your own way of displaying a UI installer in the first place.

There's few you can get wrong here.

**Comment by testbird** [comment]
> The easiest way I can think of to allow for providing both CLI and UI would be to extract the install_ methods into a sourceable script,...

That would probably mean one script per CLI and UI backend.

The other way around there already exists a sourceable bash library that allows a script to work with all the different (G)UIs: https://github.com/BashGui/easybashgui

**Comment by DrHankM** [comment]
https://github.com/BashGui/easybashgui
